### ISSUE TITLE
feat: flag private module bindings used by only one function

### ISSUE THREAD

**Issue by wkentaro** [wkentaro/gruff#74]
## Problem

A private module binding can be promoted away from its only usage site, increasing module-level state and widening the reader's search without serving another function or module.

Labelme PR #2585 introduced this verified review correction:

```python
_QIMAGE_FORMATS: Final = {
    "RGB": QImage.Format.Format_RGB888,
    "RGBA": QImage.Format.Format_RGBA8888,
}

class BrightnessContrastDialog(QtWidgets.QDialog):
    def apply(self) -> None:
        qimage = QImage(..., _QIMAGE_FORMATS[img.mode])
```

`_QIMAGE_FORMATS` is referenced only by `apply()`, so the binding belongs inside that method under the house rule “Scope a value to its usage site.” `gruff check labelme/_widgets/brightness_contrast_dialog.py` passes, and Ruff with `--select ALL` emits no scope-related finding for the reduced example.

Source: https://github.com/wkentaro/labelme/pull/2585

## Desired behavior

Gruff should deterministically flag a private module-level binding when every reference is contained in exactly one function or method and suggest moving the binding to that usage site.

A candidate diagnostic is: `Private module binding _QIMAGE_FORMATS is used by only apply; move it to that function`.

## Boundaries to settle

- Do not flag public bindings, exported names, or bindings referenced by more than one definition.
- Count references from decorators, defaults, annotations, class bodies, and module-level statements so those bindings remain at module scope.
- Decide whether functions nested inside the sole consuming definition count as the same usage site.
- Check Ruff and Pylint overlap before assigning a GR code.

## Promotion evidence

This issue records one verified review correction. Per Gruff's promotion gate, implementation should wait for repeated corrections or repository findings plus a reference-repository trial demonstrating that the deterministic shape produces review-blocking findings rather than intentional module constants.

**Comment by wkentaro** [comment]
> *This was generated by AI during triage.*

## Agent Brief

**Category:** enhancement

**Summary:** Run a reference-repository trial of the single-consumer non-public module binding shape and report whether it should ship as a rule. Deliverable is a report comment on this issue, not a rule.

**Current behavior:**

No gruff rule inspects binding scope; GR001–GR010 cover inputs, `__all__`, `Final`, docstrings, comments, tests, guards, and branch polarity. Verified on the reduced example from the issue body: gruff `--select GR` emits only GR005 for the positional-or-keyword input, Ruff `--select ALL --preview` emits only INP001, D100, CPY001, D101, D102, PLR6301, and Pylint emits only R0903. Nothing scope-related from any tool, so a GR code has no overlap. Gruff has no semantic model dependency; every existing rule is syntactic, so a shipped rule would need new reference-resolution capability. That cost is why this stage is a trial.

**Definitions:**

- **Non-public binding:** a module-level simple-name assignment or annotated assignment whose name starts with an underscore and does not end with one. Same boundary as the glossary's non-public definition.

- **Consumer:** a module-level `def` or a method of a module-level class. A reference anywhere lexically inside the consumer, including nested functions, lambdas, and comprehensions, counts as that consumer's reference.

- **Single-consumer binding:** a non-public binding with at least one reference, where every reference resolves to exactly one consumer.

**Desired behavior (the shape under trial):**

Report a single-consumer binding at its assignment target with the message:

```
Non-public module binding `_NAME` is used only by `consumer`; move it into that definition
```

The message does not prescribe the local spelling. Under GR004 the moved binding stays uppercase and `Final`, for example `QIMAGE_FORMATS: Final = {...}` inside the consuming method. The leading underscore may be dropped because a function-local name has no public/non-public distinction.

Do not flag a binding when any of these hold:

- the name is bound more than once at module level, or anywhere via `global`, `del`, augmented assignment, or a walrus;

- the name is bound by an `import`, `def`, `class`, `for`, `with`, `except`, or tuple unpacking;

- any reference sits outside a consumer: module-level statement, class body, decorator, default value, or annotation;

- the name appears as a string in `__all__`;

- a `TYPE_CHECKING` block, `globals()`, `vars()`, `eval`, or `exec` appears in the module;

- a local binding of the same name shadows it in any scope. Shadowing makes reference resolution ambiguous, so the trial skips the whole binding.

**Trial procedure:**

1. Write a throwaway Python script implementing the shape above. It is not gruff code and is not committed.

2. Run it over the six reference working trees: labelme, gdown, imgviz, osam, video-cli, imshow.

3. Hand-label every finding, or a random sample of 40 if there are more, as one of: review-blocking, intentional module constant, resolution error.

4. Report per-repo counts, labelled precision, and every resolution error with its cause.

**Acceptance criteria:**

- [ ] Report comment posted on this issue with the per-repo counts table and the precision figure.

- [ ] Each exemption above is either confirmed necessary by a labelled example or marked as unexercised in the trial.

- [ ] The nested-function question in the issue body has a recommendation backed by at least one labelled example.

- [ ] A recommendation to ship as GR011, narrow the shape, or drop, with a one-paragraph ADR-style rationale.

- [ ] The report names the open maintainer decision: whether precision must clear the 50-finding gate used for GR004, or whether an owner-taste adoption like GR010 is acceptable.

**Out of scope:**

- Adding a rule, rule doc, ADR, changelog fragment, or a semantic-model dependency.

- Public bindings, class attributes, and bindings inside functions.

- Autofix design.

- Bindings with zero consumers. Dead module bindings are a different policy.

**Source evidence note:** labelme PR #2585 is still open and its head still carries `_QIMAGE_FORMATS` at module level with one reference; the correction described in the issue body is not yet visible on the forge. Treat this as one data point.

**Comment by wkentaro** [comment]
> *This was generated by AI during implementation.*

## Trial report: single-consumer non-public module bindings

Throwaway script (in the collapsed block at the end) implementing the brief's shape, run over the six reference working trees at their current `main` heads: labelme `e4688242`, gdown `c7ebbae`, imgviz `2dd2914`, osam `5f238e0`, video-cli `17f648d`, imshow `1510ef7`.

All 59 findings were hand-labelled rather than a 40-finding sample.

### Per-repo counts

| Repo | Python files | Findings | Review-blocking | Intentional module constant | Resolution error |
|---|---|---|---|---|---|
| labelme | 145 | 51 | 47 | 3 | 1 |
| imgviz | 123 | 7 | 7 | 0 | 0 |
| osam | 47 | 1 | 0 | 1 | 0 |
| gdown | 17 | 0 | 0 | 0 | 0 |
| video-cli | 17 | 0 | 0 | 0 | 0 |
| imshow | 12 | 0 | 0 | 0 | 0 |
| **Total** | 361 | **59** | **54** | **4** | **1** |

**Precision of the shape as briefed: 54/59 = 91.5%.**

The 54 review-blocking bindings belong to 37 distinct consumers, because several sit in sibling blocks owned by one function: the eight `_EXIF_ORIENTATION_*` values in `labelme/_utils/image.py` all belong to `apply_exif_orientation`, the five `_*_POINT_COUNT` values in `examples/utils.py` all belong to `shape_to_mask`. A finding per binding is the right granularity for a linter, but a reviewer would make one correction per block.

Sensitivity: five of the 54 are the `_here = pathlib.Path(__file__).parent` idiom in imgviz's `data/*/__init__.py` loaders. Each has one consumer and `__file__` is available inside the function, so the house rule moves them, but if a maintainer treats module-location as a module property the figure is 49/59 = 83.1%.

### Labelled examples

**Review-blocking (54).** The reduced example from the issue body recurs almost verbatim: `_DISPLAYABLE_MODES: Final = {"1", "L", "P", ...}` in `labelme/_label_file.py:406` is consumed only by `_imread` three lines below it, and `_SCHEME_BY_THEME: Final = {...}` in `labelme/_utils/qt.py:18` only by `apply_color_theme`. UI literals used once in a widget `__init__` are the most common form: `_GROUP_ID_PLACEHOLDER`, `_DESCRIPTION_PLACEHOLDER`, `_LABEL_LIST_HEIGHT` in `label_dialog.py`; `_OBJECT_NAME_SUFFIX`, `_FONT_SCALE_FACTOR`, `_VERTICAL_SEPARATOR_STYLE` in `tool_bar.py`; `_ZOOM_LEVEL_LABEL` in `zoom_widget.py`. Test modules contribute 13 findings of the same shape, for example `_TOLERANCE_PX` in `tests/e2e/window_geometry_persistence_test.py` used by one test and `_STATUS_MESSAGE_TIMEOUT_MS` used by one helper.

**Intentional module constant (4).** Every one is a memoised call result whose consumer runs repeatedly, so the suggested move changes runtime cost rather than just scope:

- `labelme/_config/_writer.py:20` `_DEFAULT_CONFIG = _yaml.safe_load((here / "default_config.yaml").read_text(...))`, consumed by `_default_value`, which is called per config key. Moving it re-reads and re-parses the YAML on every call.
- `labelme/_shape.py:208` `_ORIENTED_RECTANGLE_ARROW_TEMPLATE = np.array([...]) * _ARROW_HALF_LENGTH`, consumed by `oriented_rectangle_arrow_points`, which runs on every paint of an oriented rectangle.
- `labelme/_shape_color.py:9` `_LABEL_COLORMAP = imgviz.label_colormap()`, consumed by `resolve_shape_color`, called per shape.
- `osam/_models/yoloworld/clip/clip.py:7` `_tokenizer = SimpleTokenizer()`, consumed by `tokenize`; the constructor reads and parses a gzipped BPE vocabulary.

**Resolution error (1).** `labelme/__main__.py:30` `_LOGGER_LEVELS` has one consumer inside the module (`main`), but `tests/unit/__main___test.py:16` does `from labelme.__main__ import _LOGGER_LEVELS` and parametrizes on it. Cause: a single-module walk cannot see a cross-module import of a non-public name. Ruff's `PLC2701` (`import-private-name`) flags exactly that import, so the residual error is bounded by a rule the same configuration can enable; the rule doc's suppression section should name it.

### Exemption audit

| Exemption in the brief | Status | Evidence |
|---|---|---|
| Bound more than once at module level | Unexercised in the corpus | Only hit by the synthetic smoke module |
| `global`, `del`, augmented assignment, walrus | **Confirmed necessary** | `osam/apis.py:74` declares `global _non_maximum_suppression_inference_session` inside its single consumer to lazily build and cache an ONNX session; moving the binding into the function would destroy the cache |
| Bound by `import`, `def`, `class`, `for`, `with`, `except`, unpacking | Unexercised as an exemption | Excluded by construction, since candidates are simple `Assign`/`AnnAssign` only |
| Reference at module level | **Confirmed necessary** | `labelme/_shape.py:206` `_ARROW_HEAD_BACK_OFFSET` is read by the module-level expression that builds `_ORIENTED_RECTANGLE_ARROW_TEMPLATE` |
| Reference in a decorator | **Confirmed necessary** | `labelme/tests/unit/_config/schema_test.py:33` `_ENUM_SETTINGS` is read in `@pytest.mark.parametrize(...)` |
| Reference in a default value | **Confirmed necessary** | `labelme/_widgets/label_dialog.py:45` `text: str = _PLACEHOLDER_TEXT` |
| Reference in an annotation | **Confirmed necessary** | `gdown/extractall.py:8` `_TarReadMode = Literal[...]` is used as a parameter annotation of `_extractall_tar` |
| Reference in a class body | Unexercised in the corpus | Smoke module only |
| Name appears as a string in `__all__` | Unexercised in the corpus | Smoke module only |
| `TYPE_CHECKING` block in the module | Exercised, hid no finding | Skipped `labelme/tests/e2e/ai_text_to_annotation_test.py` (its candidate `_AI_TEXT_MODEL` has three consumers anyway) and `imgviz/_nchannel.py` (no candidates) |
| `globals()`, `vars()`, `eval`, `exec` in the module | Unexercised as builtin names | Matching `exec` as an attribute too (Qt's `.exec()`) skipped 7 labelme modules and hid 8 findings, 7 of them review-blocking. The exemption must match the builtin name, not attribute access |
| Local binding of the same name in any scope | Unexercised in the corpus | Smoke module only |

### Nested-function question

Four findings reach the binding only through a nested scope inside the sole consumer, all comprehensions: `_CHANNEL_VALUE_MAX` in `labelme/_config/_shape_color.py` (generator inside `_is_rgb`), `_POINT_COORDINATE_COUNT` in `labelme/_label_file.py` (generator inside `_load_shape_json_obj`), `_EXT_BUTTON_NAME` in `labelme/tests/unit/widgets/tool_bar_test.py` (list comprehension inside `_user_buttons`), and `_tokenizer` in osam (list comprehension inside `tokenize`). Comprehensions are real scopes in Python 3, so a rule that treated nested scopes as separate usage sites would exempt three plain single-function constants that a reviewer would move. **Recommendation: a reference anywhere lexically inside the consumer counts as that consumer's reference, as the brief defines.** The moved binding stays visible from any nested scope through closure, so the same argument covers nested `def` and `lambda`, but no instance of either appeared in the corpus and that arm is unexercised.

### Narrowing that removes every false positive

All four intentional module constants have a `Call` node somewhere in their value expression, and none of the 54 review-blocking findings needs a call to be review-blocking. Restricting the shape to bindings whose value contains no call expression drops the four intentional constants and nine review-blocking ones (`os.path.join(...)`, `QColor(...)`, `QSize(...)`, `Path(__file__)...` and the five `_here` idioms), leaving 46 findings: 45 review-blocking, 0 intentional, 1 resolution error. **Precision of the narrowed shape: 45/46 = 97.8%.** The narrowing also settles the `_here` sensitivity by exempting it, and it is decidable syntactically.

### Recommendation

**Ship as GR011 in the narrowed shape** (non-public, simple-name, module-level binding whose value contains no call, bound exactly once, referenced only from inside exactly one module-level function or method, with the brief's exemptions).

ADR-style rationale: A private module binding read by one function widens every reader's search to the module and grows module state without serving another consumer; the house rule scopes a value to its usage site, and this shape identifies that case as a syntactic fact. The trial over six repositories measured 91.5% precision for the shape as briefed and 97.8% once the value is required to be call-free, because every intentional module constant found was a memoised call result whose consumer runs repeatedly, and moving such a binding changes runtime cost rather than scope. The call-free gate is chosen over a runtime-cost heuristic because it is decidable from syntax alone, keeps every review-blocking literal-valued finding, and makes the one remaining false-positive class, a non-public name imported by another module, a case that Ruff `PLC2701` already flags. Nested scopes inside the consumer count as the same usage site because comprehensions were the only nested form observed and treating them as separate sites exempts plain single-function constants. The rule needs a single-module lexical scope walk (bindings and loads per name), which is new for gruff but does not require a semantic model or cross-module resolution given the exemptions; the throwaway script's walk is about 150 lines. It remains opt-in while repository trials establish that the move reliably reduces review work.

### Open maintainer decision

The brief asks whether precision must clear the 50-finding gate used for GR004 or whether an owner-taste adoption like GR010 is acceptable. The shape as briefed produces 59 findings at 91.5%, above the 50-finding count. The narrowed shape recommended above produces 46 findings at 97.8%, below the 50-finding count. The maintainer has to decide which of these the gate measures: if the count must be at least 50, either ship the un-narrowed shape and accept the four memoisation false positives, or extend the corpus before shipping the narrowed one. If GR010-style owner-taste adoption is acceptable, the narrowed shape can ship on this trial as is.

<details>
<summary>Trial script (throwaway, not committed)</summary>

```python
"""Throwaway trial for gruff issue #74: single-consumer non-public module bindings."""
import ast, sys, json, subprocess, os, collections

MODULE_SKIP_NAMES = {"TYPE_CHECKING", "globals", "vars", "eval", "exec"}

def is_nonpublic(name):
    return name.startswith("_") and not name.endswith("_")

class Scope:
    def __init__(self, kind, name, node, parent):
        self.kind, self.name, self.node, self.parent = kind, name, node, parent

def consumer_of(scope):
    """Return the consumer (module-level def or method of module-level class) enclosing scope, or None."""
    s = scope
    chain = []
    while s is not None:
        chain.append(s)
        s = s.parent
    chain.reverse()  # module first
    # chain[0] is module
    if len(chain) < 2:
        return None
    if chain[1].kind == "func":
        return chain[1]
    if chain[1].kind == "class" and len(chain) >= 3 and chain[2].kind == "func":
        return chain[2]
    return None

def analyze(path, src):
    try:
        tree = ast.parse(src, path)
    except SyntaxError as e:
        return {"skip_module": f"syntax error: {e}"}, []
    names_in_module = {n.id for n in ast.walk(tree) if isinstance(n, ast.Name)}
    attrs_in_module = {n.attr for n in ast.walk(tree) if isinstance(n, ast.Attribute)}
    hit = (names_in_module | (attrs_in_module if os.environ.get("ATTRS") else set())) & MODULE_SKIP_NAMES
    if hit:
        return {"skip_module": "module uses " + ",".join(sorted(hit))}, []

    # candidate bindings: module-level simple Assign / AnnAssign with value
    candidates = {}
    for stmt in tree.body:
        if isinstance(stmt, ast.Assign) and len(stmt.targets) == 1 and isinstance(stmt.targets[0], ast.Name):
            candidates.setdefault(stmt.targets[0].id, []).append(stmt)
        elif isinstance(stmt, ast.AnnAssign) and isinstance(stmt.target, ast.Name) and stmt.value is not None:
            candidates.setdefault(stmt.target.id, []).append(stmt)
    candidates = {k: v for k, v in candidates.items() if is_nonpublic(k)}
    if not candidates:
        return {}, []

    # __all__ strings
    all_strings = set()
    for n in ast.walk(tree):
        if isinstance(n, (ast.Assign, ast.AnnAssign, ast.AugAssign)):
            tgts = n.targets if isinstance(n, ast.Assign) else [n.target]
            if any(isinstance(t, ast.Name) and t.id == "__all__" for t in tgts):
                for c in ast.walk(n):
                    if isinstance(c, ast.Constant) and isinstance(c.value, str):
                        all_strings.add(c.value)

    # Walk with scopes: collect bindings (all kinds) and loads per name.
    binds = collections.defaultdict(list)   # name -> [(scope, kind, node)]
    loads = collections.defaultdict(list)   # name -> [(scope, region, node)]
    # region: "body" (inside a consumer body), or "decorator"/"default"/"annotation"/"classbody"/"module"/"nested-class-method"/...

    def bind(name, scope, kind, node):
        binds[name].append((scope, kind, node))

    def visit(node, scope, region):
        # region describes where a load sits relative to a consumer
        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            bind(node.name, scope, "def", node)
            for d in node.decorator_list:
                visit(d, scope, "decorator")
            a = node.args
            for dflt in a.defaults + [d for d in a.kw_defaults if d is not None]:
                visit(dflt, scope, "default")
            for arg in a.posonlyargs + a.args + a.kwonlyargs + ([a.vararg] if a.vararg else []) + ([a.kwarg] if a.kwarg else []):
                if arg.annotation is not None:
                    visit(arg.annotation, scope, "annotation")
            if node.returns is not None:
                visit(node.returns, scope, "annotation")
            inner = Scope("func", node.name, node, scope)
            for arg in a.posonlyargs + a.args + a.kwonlyargs + ([a.vararg] if a.vararg else []) + ([a.kwarg] if a.kwarg else []):
                bind(arg.arg, inner, "arg", arg)
            for s in node.body:
                visit(s, inner, "body")
            return
        if isinstance(node, ast.Lambda):
            a = node.args
            for dflt in a.defaults + [d for d in a.kw_defaults if d is not None]:
                visit(dflt, scope, region if region == "body" else "default")
            inner = Scope("lambda", "<lambda>", node, scope)
            for arg in a.posonlyargs + a.args + a.kwonlyargs + ([a.vararg] if a.vararg else []) + ([a.kwarg] if a.kwarg else []):
                bind(arg.arg, inner, "arg", arg)
            visit(node.body, inner, region)
            return
        if isinstance(node, ast.ClassDef):
            bind(node.name, scope, "class", node)
            for d in node.decorator_list:
                visit(d, scope, "decorator")
            for b in node.bases + [k.value for k in node.keywords]:
                visit(b, scope, "classbase")
            inner = Scope("class", node.name, node, scope)
            for s in node.body:
                visit(s, inner, "classbody" if region != "body" else "body")
            return
        if isinstance(node, (ast.ListComp, ast.SetComp, ast.GeneratorExp, ast.DictComp)):
            inner = Scope("comp", "<comp>", node, scope)
            first = True
            for gen in node.generators:
                visit(gen.iter, scope if first else inner, region)
                first = False
                for t in ast.walk(gen.target):
                    if isinstance(t, ast.Name):
                        bind(t.id, inner, "comp-target", t)
                for cond in gen.ifs:
                    visit(cond, inner, region)
            if isinstance(node, ast.DictComp):
                visit(node.key, inner, region); visit(node.value, inner, region)
            else:
                visit(node.elt, inner, region)
            return
        if isinstance(node, ast.Global):
            for n in node.names: bind(n, scope, "global", node)
            return
        if isinstance(node, ast.Nonlocal):
            for n in node.names: bind(n, scope, "nonlocal", node)
            return
        if isinstance(node, (ast.Import, ast.ImportFrom)):
            for al in node.names:
                nm = (al.asname or al.name).split(".")[0]
                bind(nm, scope, "import", node)
            return
        if isinstance(node, ast.ExceptHandler) and node.name:
            bind(node.name, scope, "except", node)
        if isinstance(node, ast.MatchAs) and node.name:
            bind(node.name, scope, "match", node)
        if isinstance(node, ast.MatchStar) and node.name:
            bind(node.name, scope, "match", node)
        if isinstance(node, ast.MatchMapping) and node.rest:
            bind(node.rest, scope, "match", node)
        if isinstance(node, ast.Name):
            if isinstance(node.ctx, ast.Load):
                loads[node.id].append((scope, region, node))
            elif isinstance(node.ctx, ast.Store):
                kind = "store"
                bind(node.id, scope, kind, node)
            else:
                bind(node.id, scope, "del", node)
            return
        if isinstance(node, ast.AugAssign):
            # target is Store ctx; mark specially
            if isinstance(node.target, ast.Name):
                bind(node.target.id, scope, "augassign", node)
                visit(node.value, scope, region)
                return
        if isinstance(node, ast.NamedExpr):
            bind(node.target.id, scope, "walrus", node)
            visit(node.value, scope, region)
            return
        if isinstance(node, (ast.For, ast.AsyncFor)):
            for t in ast.walk(node.target):
                if isinstance(t, ast.Name): bind(t.id, scope, "for", t)
            visit(node.iter, scope, region)
            for s in node.body + node.orelse: visit(s, scope, region)
            return
        if isinstance(node, ast.withitem) and node.optional_vars is not None:
            for t in ast.walk(node.optional_vars):
                if isinstance(t, ast.Name): bind(t.id, scope, "with", t)
            visit(node.context_expr, scope, region)
            return
        if isinstance(node, ast.Assign):
            for t in node.targets:
                if isinstance(t, ast.Name):
                    bind(t.id, scope, "assign", t)
                elif isinstance(t, (ast.Tuple, ast.List)):
                    for n in ast.walk(t):
                        if isinstance(n, ast.Name) and isinstance(n.ctx, ast.Store): bind(n.id, scope, "unpack", n)
                        elif isinstance(n, ast.Name): loads[n.id].append((scope, region, n))
                        elif isinstance(n, (ast.Attribute, ast.Subscript)):
                            pass
                    # loads inside attribute/subscript targets
                    for n in ast.walk(t):
                        if isinstance(n, (ast.Attribute, ast.Subscript)):
                            visit(n.value if isinstance(n, ast.Attribute) else n.value, scope, region)
                            if isinstance(n, ast.Subscript): visit(n.slice, scope, region)
                else:
                    visit(t, scope, region)
            visit(node.value, scope, region)
            return
        if isinstance(node, ast.AnnAssign):
            if isinstance(node.target, ast.Name):
                bind(node.target.id, scope, "assign" if node.value is not None else "annonly", node.target)
            else:
                visit(node.target, scope, region)
            visit(node.annotation, scope, region if region == "body" else "annotation")
            if node.value is not None: visit(node.value, scope, region)
            return
        for child in ast.iter_child_nodes(node):
            visit(child, scope, region)

    mod = Scope("module", "<module>", tree, None)
    for s in tree.body:
        visit(s, mod, "module")

    findings, exemptions = [], collections.Counter()
    for name, stmts in candidates.items():
        b = binds[name]
        mod_binds = [x for x in b if x[0] is mod]
        other_binds = [x for x in b if x[0] is not mod]
        kinds = collections.Counter(k for _, k, _ in b)
        if name in all_strings:
            exemptions["in __all__"] += 1; continue
        if len(mod_binds) != 1:
            exemptions["bound more than once at module level (%s)" % ",".join(sorted(set(k for _,k,_ in mod_binds)))] += 1; continue
        if any(k in ("global","augassign","walrus","del","nonlocal") for _, k, _ in b):
            exemptions["global/del/augassign/walrus"] += 1; continue
        if other_binds:
            exemptions["shadowed in inner scope (%s)" % ",".join(sorted(set(k for _,k,_ in other_binds)))] += 1; continue
        refs = loads[name]
        if not refs:
            exemptions["zero references (out of scope)"] += 1; continue
        outside = [(r, sc) for sc, r, n in refs if r != "body" or consumer_of(sc) is None]
        if outside:
            regions = sorted(set(r if r != "body" else "nested-class-method" for r, sc in outside))
            exemptions["reference outside a consumer (%s)" % ",".join(regions)] += 1; continue
        consumers = {id(consumer_of(sc)): consumer_of(sc) for sc, r, n in refs}
        if len(consumers) != 1:
            exemptions["multiple consumers"] += 1; continue
        c = next(iter(consumers.values()))
        cname = c.name if c.parent.kind == "module" else f"{c.parent.name}.{c.name}"
        nested = any(sc is not c for sc, r, n in refs)
        nested_kinds = sorted(set(sc.kind for sc, r, n in refs if sc is not c))
        stmt = stmts[0]
        findings.append({
            "file": path, "line": stmt.lineno, "name": name, "consumer": cname,
            "consumer_line": c.node.lineno,
            "refs": [n.lineno for sc, r, n in refs],
            "via_nested": nested_kinds,
            "value": ast.get_source_segment(src, stmt.value)[:120].replace("\n", " "),
            "value_lines": (stmt.end_lineno - stmt.lineno + 1),
        })
    return dict(exemptions), findings

def main(repos):
    out = {}
    for repo in repos:
        files = subprocess.run(["git", "-C", repo, "ls-files", "*.py", "**/*.py"], capture_output=True, text=True).stdout.split()
        repo_ex, repo_f, module_skips = collections.Counter(), [], collections.Counter()
        for f in files:
            p = os.path.join(repo, f)
            try:
                src = open(p, encoding="utf-8").read()
            except Exception:
                continue
            ex, fs = analyze(f, src)
            if "skip_module" in ex:
                # count only if module actually has candidates
                module_skips[ex["skip_module"]] += 1
                continue
            repo_ex.update(ex); repo_f.extend(fs)
        out[os.path.basename(repo)] = {"files": len(files), "findings": repo_f, "exemptions": dict(repo_ex), "module_skips": dict(module_skips)}
    json.dump(out, sys.stdout, indent=1)

if __name__ == "__main__":
    main(sys.argv[1:])
```

</details>
